### ISSUE TITLE
ember-auto-import only supports dynamic import() with a string literal argument

### ISSUE THREAD

**Issue by bjornharrtell** [ef4/ember-auto-import#97]
This is of course an expected error due to https://github.com/ef4/ember-auto-import/blob/f36a44028e2fa18378a9a7d9b646ae5a7d944c8d/ts/analyzer.ts#L148 but I fail to see the rationale of it.

If I use ember-auto-import version 1.1.0 I can use dynamic import with runtime variable generated argument without any apparent issues (assuming a browser that can handle it).

Indeed I fail to understand what the purpose of the support for dynamic import in 1.2.0 is. Regardless, would it make sense to make it optional?

**Comment by ef4** [comment]
> without any apparent issues (assuming a browser that can handle it).

There is another assumption lurking in there: that the user has somehow manually produced Javascript files that will satisfy whatever runtime resolution is required. That is a big assumption, and it's contrary to what ember-auto-import is doing the rest of the time.

Most of the time, saying `import "something"` _causes_ something to get into the final build output. And the same holds true for `import("something")` -- that _causes_ it to get packaged into its own chunk that will definitely be resolvable at runtime.

So my concern is around not creating a trap for people. It's surprising if `import("something")` causes "something" to magically be included in the build, but `import((function(){ return "something" })())` silently does not, and results in either a parse error or a rejected promise at runtime (depending on whether the browser has `import()` support).

> Regardless, would it make sense to make it optional?

That's not really one of the choices -- even if your code doesn't expect ember-auto-import to do anything with it, your dependencies may, or their dependencies. We must have a policy on what to do when we encounter it. And if we simply ignore it, people will get broken builds. I have already had real bug reports about this, it does happen.

I'm not against relaxing the current restriction, which I agree is extreme and was only intended as the starting point. But _some_ restriction will be needed. For example, we could allow template strings that start with a literal part, like `import('./lang/${language}')`, which would let us narrow down the universe of possible imports to a particular directory.

**Comment by bjornharrtell** [comment]
Thanks for the thorough explanation. It was "the same holds true for import("something") -- that causes it to get packaged into its own chunk that will definitely be resolvable at runtime" that was not clear to me. I agree this might the safe and common use case and that it it's helpful to let the user know when they are venturing outside that.

However, I still feel this should be a warning instead of bail out.

In particular "user has somehow manually produced Javascript files that will satisfy whatever runtime resolution is required" is the use case I have and I fail to see that as unreasonable. In my case it's even dynamically configured, i.e some external dependencies can be configured at runtime by the user, might reside on another host entirely and as such are completely unknown at build time.

**Comment by ef4** [comment]
Yes, what we need to design is a way to allow people to opt-out of the automatic packaging _for a particular_ `import()` callsite. You want to say "leave this particular import() alone, I promise I will take care of the packaging myself".

There's nothing in the language itself that makes it easy to annotate such a spot. Webpack works around this by using actual comments like `import(/* webpack-specific options go here */ someThing)`. We _could_ do that, though I don't particularly love it.

We could also do it at the module level, like having a config option that says "don't try to package any `import()` you see in these particular files..."

Another workaround that you could use immediately is to put the `import()` call in a place that ember-auto-import isn't analyzing. For example, if you make an in-repo addon and do the import from inside there, and the addon doesn't list ember-auto-import as a `dependency`, then we don't analyze its code and we won't see the import(), and it will be left alone.

**Comment by bjornharrtell** [comment]
The break out into add-on sounds awfully a lot like the "you're on your own" opt-out option I'd like to see, however cumbersome. ;) Can I not persuade you to reconsider a setting to disable analyzing dynamic imports (perhaps selectively)? I'm willing to work on a PR for that.

**Comment by ef4** [comment]
I'm OK with a _selective_ solution. But it needs to be future-compatible, so I'd like to avoid any special settings given to ember-auto-import. 

Rather, what if we allowed this?

```js
import(`https://${customScript}`);
```

It would be allowed because we can see statically that it definitely contains `//`, so it is therefore an absolute URL. To me, that sufficiently conveys to readers that they can't expect ember-auto-import to do anything at build-time for them here. It creates a clear intent to escape the build system.

If you're concerned about multiple protocols, this could also be allowed (it follows the static `//` rule):

```js
import(`${proto}://${originAndPath}`)
```

The template strings would be mandatory here. It nicely limits the possible syntax that we would be obligated to understand. Users can put whatever complicated expressions they want into the dynamic parts, as long as they also provide the static parts that we can see.

If some of your custom URLs are actually relative, you would be obligated to make them absolute first, but that is straightforward using the [URL API](https://developer.mozilla.org/en-US/docs/Web/API/URL).

**Comment by bjornharrtell** [comment]
Ah yes, that seem like a nice way to handle it and should be easy to document.

**Comment by rwwagner90** [comment]
@ef4 How are we supposed to import dynamic things, if this is not allowed? 

I'm trying to support importing language files in ember-flatpickr, but not sure how to do it. I was trying something like this:

```js
const localeString = `flatpickr/dist/l10n/${newLocale}.js`;
          import(localeString).then((localeModule) => {
            this.field._flatpickr.set('locale', localeModule);
          });
```

**Comment by ef4** [comment]
@rwwagner90 this original issue was asking a slightly different thing (how to get ember-auto-import to ignore certain dynamic imports entirely, rather than how to get a "wildcard" import to get properly packaged so that every valid alternative is available at runtime). 

But I wrote up the current state of your question here: https://github.com/embroider-build/embroider/issues/198

It's on the embroider repo, because ember-auto-import is intended as a polyfill for embroider and will follow the same spec.

**Comment by brunoocasali** [comment]
@rwwagner90 how do you achieve this behaviour? 

I'm trying to create a data-loader component using `ember-graphql`, and I can't request data to server without passing a imported graphql file.

```js
// this exists => 'backstage/gql/queries/activities' 

// component.js
const name = `backstage/gql/queries/${this.args.query}`;
const query = yield import(name).then(...)

this.results = yield this.apollo.query({ query, variables }, this.args.query)
```

**Comment by rwwagner90** [comment]
@brunoocasali see https://github.com/ef4/ember-auto-import/issues/97#issuecomment-496645346 TL;DR you cannot do it.

**Comment by rwwagner90** [comment]
@ef4 this is sorely needed. I still do not understand why we cannot dynamically build a string to import. Can we please support this?

**Comment by ef4** [comment]
Yes, we have consensus on the plan for this that’s included in the embroider RFC.

There’s an open PR here with implementation. https://github.com/ef4/ember-auto-import/pull/293

**Comment by rwwagner90** [comment]
Ah awesome, thanks!

**Comment by ef4** [comment]
This was fixed back in 1.10.0.
